contracts/GroupVoteLocker.sol

```diff
@@ -0,0 +1,331 @@
+            }
+        } else if (checkpoint0.blk != block.number) {
+            /* to estimate how much time has passed since the last checkpoint get the number
+             * of blocks since the last checkpoint. And multiply that by the average time per 
+             * block since the last checkpoint and present blockchain state. 
+             */
+            dTime =
+                ((_blockNumber - checkpoint0.blk) *
+                    (block.timestamp - checkpoint0.ts)) /
+                (block.number - checkpoint0.blk);
+        }
+        // if code doesn't enter any of the above if conditions latest _blockNumber was passed
+        // to the function and dTime is correctly set to 0
+
+        // Now dTime contains info on how far are we beyond point
+        return _votePowerAt(groupState, checkpoint0, checkpoint0.ts + dTime);
+    }
+
+    /**
+     * @dev Calculates voting power at a given time _t
+     * @param groupState group state to calculate voting power from
+     * @param _checkpoint Most recent point before time _t
+     * @param _time Time at which to calculate supply
+     * @return totalSupply at given time
+     */
+    function _votePowerAt(GroupVotePowerState storage groupState, Checkpoint memory _checkpoint, uint256 _time)
+        internal
+        view
+        returns (uint256)
+    {
+        Checkpoint memory lastCheckpoint = _checkpoint;
+
+        // Floor the timestamp to weekly interval
+        uint256 iterativeTime = _floorToWeek(lastCheckpoint.ts);
+        // Iterate through all weeks between _checkpoint & _time to account for slope changes
+        for (uint256 i = 0; i < 255; i++) {
+            iterativeTime = iterativeTime + WEEK;
+            int128 dSlope = 0;
+            // If week end is after timestamp, then truncate & leave dSlope to 0
+            if (iterativeTime > _time) {
+                iterativeTime = _time;
+            }
+            // else get most recent slope change
+            else {
+                dSlope = groupState.slopeChanges[iterativeTime];
+            }
+
+            lastCheckpoint.bias =
+                lastCheckpoint.bias -
+                (lastCheckpoint.slope *
+                    SafeCast.toInt128(
+                        int256(iterativeTime - lastCheckpoint.ts)
+                    ));
+
+            if (iterativeTime == _time) {
+                break;
+            }
+
+            lastCheckpoint.slope = lastCheckpoint.slope + dSlope;
+            lastCheckpoint.ts = iterativeTime;
+        }
+
+        return SafeCast.toUint256(max(lastCheckpoint.bias, 0));
+    }
+
+    /**
+     * @dev Binary search (bisection) to find epoch closest to block.
+     * @param _block Find the most recent point history before this block
+     * @param _maxEpoch Maximum epoch
+     * @return uint256 The most recent epoch before the block
+     */
+    function _findEpoch(
+        Checkpoint[] memory _checkpoints,
```

Comment:
_findEpoch is designed to efficiently search inside a list of checkpoints. However, rather than searching storage slots, it works from memory. This means that before it can begin searching, it first *loads every single checkpoint struct from storage*.

Not only does it have to read every storage slot before being called, it's reading all the fields of these structs in, even the ones that won't be looked at.

This way of working probably needs a change.

We need to make sure we have unit tests that work on 400 checkpoint systems.

Reply:
> 🟥 Given that anyone can create as many global snapshots as they want per block, and this has a linear number of storage reads per snapshot, this could be a contract killer here.

The loop is bounded so I think in the worst case you'll end up with it finding the wrong epoch rather than a broken contract. The consequences of that are that the historical total supply will break.

Exponential increases in checkpoint values only result in linear increases in gas cost here, so I think it's probably unfeasible to use this as an attack vector. We should write some tests though.

Reply:
@tomlinton The problem isn't the loop - the problem is the `memory` attribute here. This will load EVERY element in from storage before running the function. It's an implicit loop over every checkpoint ever stored in this array.

Reply:
Ahh I see what you mean 👍